### DISCUSSION TITLE
Spacy support

### DISCUSSION THREAD

**Discussion by Vilhelm-Ian** [mortii/anki-morphs#13] [Ideas]
For languages without spaces like german there is no dictionary So K and V are same. So morphman gives me the same word different endings. It would be nice if morphman used Spacy

**Comment by mortii** [comment]
Thanks for the suggestion! Could you provide some additional info:
1. Could you give me a link to Spacy?
2. Can you give some examples of words that should be treated differently?

**Comment by CodeWithMa** [comment]
1. https://spacy.io/
2. Take a look at this issue: https://github.com/kaegi/MorphMan/issues/162

I think the latest PR was this one: https://github.com/kaegi/MorphMan/pull/231

**Comment by mortii** [comment]
> 1. https://spacy.io/
> 
>     2. Take a look at this issue: [Add spaCy support for MorphMan kaegi/MorphMan#162](https://github.com/kaegi/MorphMan/issues/162)
> 
> 
> I think the latest PR was this one: [kaegi/MorphMan#231](https://github.com/kaegi/MorphMan/pull/231)

OMG! That is absolutely amazing! My mind is blown, this will take the project to a completely new level! Thank you for bringing it to my attention, this will now be on top of the todo-list after the refactoring is done!

**Comment by ashprice** [comment]
Of interest as well might be https://github.com/megagonlabs/ginza

**Reply by Vilhelm-Ian** [reply]
AnkiMorphs has a dictionary for japanese. I think using a NLP library would give worse results

**Reply by ashprice** [reply]
This turned into a wall of text, and I am very sleep deprived, but below is basically the kind of idea I had in mind for Ginza. Actually implementing this though? It seems like a headache and I don't know if it'd be significantly better for language learning purposes than just using sudachidict with sudachi's mode C tokenizer for spacy. 

The default spacy model is smaller than the ginza model and trained to work best with sudachipy's 'A' mode tokenizer. This has the disadvantage of splitting, for example, compound verbs. Nouns for the most part are treated in the way we'd want compound verbs to be treated - this is probably the opposite to what you want as a learner: compound nouns are, on average, far less idiosyncratic than verbal compounds. If you are learning English and you know _eye_ and you know _doctor_, you know _eye doctor_, but if you know _under_ and you know _stand_, you probably will not understand the word _understand_ at first sight. Many Japanese compound verbs are the exact same deal (many are easy to understand, equally!)

Also edited note: by compound verbs I mean verbs of two lemmas, not verbs followed by any number of auxiliary verbs. But of course, ultimately, Ginza provides enough analytical information that _ultimately_ you could sort input based on the complexity of the internal grammatical structure of the auxiliary verbal phrase too. (But that would kind of be the far end of the roadmap, an idealised goal, not something to implement first.)

Ginza on the other hand is trained with a larger dataset (I don't think the licensing should be an issue unless it's decided to monetize this addon) and is explicitly built around leveraging sudachipy's 'C' mode tokenizer, which means that:

1. Things like compound verbs are recognized well and easily given to us in a useful form. This is great because a given compound verb may have an idiosyncratic or non-compositional meaning, or one that appears that way for a non-fluent learner. 

This extends to other constructions beyond compound verbs because of how Ginza analyzes the parse tree given to it by sudachipy - ie. whereas the typical models are best-suited to returning 'lemmas' in a very specific sense, by leveraging Ginza and some data from some 国語辞典, you can sort the Anki cards based on larger-scale lexical structures which is _far_ more likely to (a) allow for accurate readability analysis and (b) allow for better sorting for cards that have Known+1Unknown. 

2. Leveraging 国語辞典 entries on Ginza's dependency info would allow for easier construction of an approach that incorporates lexicographical definitions into Anki cards. With base sudachipy + spacy, if you wanted to fetch definitions automatically, you're going to have to match for the unknown lemmas where those unknown lemmas are the shortest possible tokenised units - they may not even be the head of the lexical item they are occuring in! (I personally have 0 interest in an automatic-definitions feature, but I can imagine it would generate interest by others - see the popularity of vocab cards or the card structure where you have a sentence + target word (maybe clozed in the sentence) + a definition of that word on a card. There are many anki decks and commercial SRS systems using this format.) 

---

Of course you can do the good old approach of concatenating stuff together and querying a dictionary for the longest returned item. This is error prone because you have to figure out ways to make sure you have the right concatenation based on info provided by the parse tree. The idea is to get Ginza to do the heuristics for you. 

---

A TLDR and another way of phrasing all this is that mecab may mislead you in terms of what to expect with regard to tokenizations from the out of the box spacy output. Spacy by default is going to treat larger-than-one-[root](https://en.wikipedia.org/wiki/Root_(linguistics)) lexical units as separate units, and hence AnkiMorphs/Morphman et al. if using spacy-as-default would then sort cards based on 1 element within a lexeme being unknown, as opposed to whether the whole lexical item is unknown. Of course you can instead of using Ginza use spacy with sudachi in mode C, which is what Ginza is also doing, but the base spacy models are intended to be used with mode A. 

Ultimately I could see this leading down a path where you can start to dynamically sort cards based on grammatical structures as well as lexical items - Ginza gives all (or most of) the information needed to do so, for sure. Implementing that, though, is a lot of legwork. 

I'd see that as a kind of natural progression - going from being able to sort cards using oft-incorrect lemmas/lexemes (mecab) -> more often correct lemmas (spacy + sudachi) -> actual lexemes (sudachi mode C or spacy + ginza paired with a database of 国語辞典 lexemes) -> being able to sort by larger phrasal grammatical constructions that can be graded on a scale of 'known' to 'unknown.' One thing that would remain is adding some heuristic to also sort based on idiomaticity - there are projects that have been done that offer some guidance there, like massif.la . The reason we don't want to use lexemes for the original tokenization (and instead use sudachidict) is because that would give us a lot of incorrect results; it's not so much of an issue once you have a good parse tree, and Ginza handles the analysis of that tree. 

This all said, you'd have to bundle the Ginza model, for example, in addition to everything else. I don't know whether all of this would be remotely worth the headache.

**Reply by Vilhelm-Ian** [reply]
@ashprice for the nouns issue Spacy can tell you if a word is a noun to lemmatize it. The problem with ginza it's japanese specific we want ankimorphs to be a general morphemizer. I did not get why you don't use mecab dictionary. And you being knowledgeable on the topic what do you think of just using https://github.com/aboSamoor/polyglot

**Reply by mortii** [reply]
@ashprice absolutely amazing insights, thank you so much! I have a feeling I'm going to reread your comment 50 times when I start really working on this.

On licensing: the current mpl-2 license is semi-arbitrary. I don't plan on ever making money of off this add-on. I'm not a super fan of the really hard copyleft licenses mainly because I want anyone to be able to fork this and do whatever they want with it.

**Comment by mortii** [comment]
Order of priority:
1. import spacy to anki by any means possible.
2. test if it gives us useful output, if yes:
    1. make it work (reasonably) well
    2. figure out a good approach for importing spacy to anki

**Comment by Vilhelm-Ian** [comment]
A different approach we can take is to generate dictionary files using spacy for the availabel models. The dictionary would be something like the frequency.csv file lemma,base . And add them to addon. A benefit for this approach is that they would be available for people to correct them since spacy dosen't always give accurate results

**Reply by mortii** [reply]
For germanic languages the dictionaries would be infinitely long no?

**Reply by mortii** [reply]
I haven't really thought about how the japanese mecab morphemizer works, maybe it could reveal some useful tips on how to implement a spacy morphemizer.

**Reply by Vilhelm-Ian** [reply]
For germanic languages the dictionaries would be infinitely long no?

We can just generate it for the most common X words and treat the rest as the base and the lemma as the same word like we currently do with the Space morphemizer. The next point dosen't bear much weight just throwing it out there I just generated a frequency file for over twenty books. The result was 80k unique words. I am guessing that after two hundred thousand would deal with only really rare words. I really like this approach since it would be a lot simpler than to figure out how to import spacy into anki. And it would be a lot faster than having spacy in my case run on 80k cards every time I recalc. 

I guess a potential solution for this is to run manually spacy when he feels like he has added enough new words. Or to first check if the word is in morphs.db and if not then to run spacy(I am brainstorming again)

**Reply by mortii** [reply]
We could just set an arbitrary cut off, true.

**Reply by mortii** [reply]
The fundamental point of the morphemizer is to split the words of an arbitrary expression right? If you only have a dictionary then we have to do all the hard work of creating a splitter... Again, i haven't looked at the mecab code, but I have a feeling that doing all that is hard and ugly.

**Reply by Vilhelm-Ian** [reply]
Yes. That's the point of the morphemizer. But the japanese morphemizer also is able to mark different forms of the word as known. The problem I have with the "Languages with spaces morphemizer" is that if I have learned the word "reading" it can still show me a card where the word "read" is marked as unknown.

The way I invision it is for languages without spaces to split the words with spaces and then check each word if it's known if it's lemma is already known .

Talking about this got me intrested I decided to search witch languages don't have spaces that are currently used:
- Chinese(already morphemizer exists)
- Japanese(already morphemizer exists)
- Amharic(but they do have a seperator it's just : instead of space)
- Bengali(not sure if true)
- Tamil
- Sinhala
- Malayalm
-  Devanagari-based scripts: Some languages written in scripts derived from Devanagari, such as Hindi, Marathi, Nepali, and Sanskrit, typically do not use spaces between words.
- Lepacha(a minor language in india)
- Thai
- And a couple of more. Couldn't check them

**Reply by mortii** [reply]
I think that approach is an excellent backup solution if we are unable to import spacy.  imo we should try importing spacy first and hopefully it can give us all of this for free :pray:

**Reply by mortii** [reply]
by that I mean hopefully spacy can generate lemma, base for us on the fly so we don't have to create an intermediary list.

**Reply by ashprice** [reply]
Spacy can tell you the lemma is fex. 'run' for run, runs, running, though for some languages it'll be better than others.  

If you're generating a frequency dictionary I guess you're probably either counting each of those as separate items (I'd kind of object to calling them _lemmas_ if that's what you're doing) or you're using something (probably inferior) to spacy to pick 'run' as the lemma for each of those. 

Unless you mean, a process like this?
1. Use spacy to generate frequency lists for some large corpora.
2. Given Anki card with input _i_, have spacy tokenize it. 
3. For each token grep some frequency csv for a form that matches that form with the highest frequency, pull out it's respective lemma (base form)
4. Use that to sort cards.

Maybe you're doing some other way that's obvious...

Anyway, that method (1-4) is likely to increase mistakes, not decrease them - you're effectively doing something the lemmatizers in spacy already do, only without the benefit of a parse tree, that is, you are taking the tokenisation and lemmatizing it yourself, without leveraging any information from the context like the model (ie. more than likely getting worse results). One way you could use such frequency data is if you prompt users for input for most likely cases, but that sounds more fiddly than users just living with the odd misparse... I'd sure as hell rather occasionally get 1 form occasionally attributed to the wrong root word because it looks the same, than the current situation with old morphman where any language with spaces just treats _every single form_ as a single '''lemma'''' _every time, all the time._

Of course, languages that lack lemmatizers, that's probably just not something to worry about right away.

**Comment by Vilhelm-Ian** [comment]
If you want to add something that split a sentence to words if the sentence dosen't contain spaces. We can use this module https://github.com/aboSamoor/polyglot I think it will be much simpler since you don't have to download the models for a specific language it comes with the module. Also it supports more languages than spacy. Spacy claims to support 75+ while this module claims 150+.

**Reply by mortii** [reply]
:exploding_head:

**Reply by mortii** [reply]
crazy good

**Reply by Vilhelm-Ian** [reply]
@mortii can you see if this will be easier to import

**Reply by mortii** [reply]
@Vilhelm-Ian I think any third-party library is going to have the same import problem

**Reply by mortii** [reply]
I just tried to experiment with polyglot and just installing it locally for myself is a pain, so I think spaCy is a better option.

**Comment by mortii** [comment]
On the issue of importing spacy (or other packages), worst case scenario we could just sub-module them, that way we have direct access to the code in ankimorphs.

**Reply by mortii** [reply]
this doesn't work unfortunately, you would have to recursively sub-module spaCy's dependencies...

**Reply by mortii** [reply]
[AnkiSpacy](https://github.com/rteabeault/AnkiSpacy ) installs spacy by using pip from inside Anki. If we use that approach then we would have to create a manager that can also uninstall the pip packages, but that's within reason I think.

AnkiSpacy is actually amazing, too bad it's outdated...

**Comment by mortii** [comment]
I've started experimenting with spacy, and I came across this:
```
w: 診
w.norm_: 見る
w.lemma_: 診る
```

So maybe we should store morphs as 3 element tuples instead of just 2.

Mecab just gives this:
```
morph: 診
morph.norm: 診る
morph.inflected: 診
```

EDIT:
lemma & surface lemma (base & inflected) are sufficient, so frequency files can keep the format they currently have.

**Reply by mortii** [reply]
Another fun example:

w: ん
w.norm_: ず
w.lemma_: ぬ

**Comment by mortii** [comment]
Recalc duration with 37K japanese cards:
- mecab: ~8 second
- spacy.lang.ja (split mode C): ~12 seconds

so performance is thankfully not an issue (on my setup at least)

**Reply by Vilhelm-Ian** [reply]
Mortii can you teach me how you packaged spacy

**Reply by mortii** [reply]
@Vilhelm-Ian

terminal:
``` bash
source venv/bin/activate
python3.9 -m pip install spacy[ja]
```

\_\_init\_\_.py:
``` python
def test_function() -> None:
    from spacy.lang.ja import Japanese

    cfg = {"split_mode": "C"}
    nlp = Japanese.from_config({"nlp": {"tokenizer": cfg}})
    expression = "ちょっと診てやってくれないか"
    doc = nlp(expression)

    for w in doc:
        print(f"w: {w}")
        print(f"w.lemma_: {w.lemma_}")

```

When running anki outside of venv it crashes of course.

**Comment by mortii** [comment]
If we have the users install spacy and the models themselves (pip install [...]) then implementing spacy would become trivial. I would prefer to have AM do that heavy lifting though, it can be complicated and intimidating if you are not very technical. 

Edit: 
I found a cool answer about importing third party libraries: https://forums.ankiweb.net/t/best-practice-for-add-ons-using-third-party-packages/1585/2

I'm going to try adding spacy and its dependencies to a libs folder to see if that is a workable approach.

**Reply by Vilhelm-Ian** [reply]
Why would it be then trivial. I thought anki dosen't provide access to system packages?

**Reply by mortii** [reply]
@Vilhelm-Ian You have access to system packages.

You can use `import spacy` in python code just fine if spacy is already installed and in the python path. It's getting it there that is the tricky part.

**Reply by mortii** [reply]
I actually managed to import spacy this way, but it has a numpy dependency, and numpy can't be used from source since it has c files (https://stackoverflow.com/questions/14570011/explain-why-numpy-should-not-be-imported-from-source-directory)

**Reply by mortii** [reply]
i think we have to go for the user installed route unfortunately....

**Reply by Vilhelm-Ian** [reply]
I am actually fine with this route. Most people will find language with spaces fine enough. And those who really really want to use spacy won't have much issue just using "pip install"

**Reply by mortii** [reply]
Bundling spacy with AM will be impossible even if we get it to work, because the sudachidict_core depencency alone is 200MB+... So user installed is the only way forward.

@Vilhelm-Ian It's a little more complicated because they would also have to install python first if they don't already have it.... but there is no other choice :man_shrugging:

**Comment by mortii** [comment]
I'm thinking the 'morphemizer' drop down menu should be replaced by a 'model' drop down that contains all the installed spacy models, e.g.  'ja_core_news_sm'

![Screenshot from 2023-12-18 16-40-52](https://github.com/mortii/anki-morphs/assets/15674619/9a8558b7-4f4d-4997-b776-9b9068db11db)

**Reply by Vilhelm-Ian** [reply]
~~I think we can just leave the the dropdown be called Morphemizer and have it list both available spacy models and morphemizers~~ I just remembered there is no option to get a list of all spacy models. I have no idea how to make it user friendly

**Reply by mortii** [reply]
@Vilhelm-Ian when you click a drop down menu called morphemizer and you see options like  'ja_core_news_sm' that would be a little confusing no?

**Reply by mortii** [reply]
the spacy cli has a command for downloading models, so it might not be impossible to create something like yous suggested @Vilhelm-Ian 

e.g.
get a list of avilable models somehow and when selected do:
```python
import spacy
spacy.cli.download('ja_core_news_sm', ...)
```

not sure if that command will work from within anki though.

EDIT:
Having thought about it some more, I have come to the conclusion we should only list installed models, less headache for us to deal with.

**Reply by mortii** [reply]
getting a list of installed models is easy:
```python
names = spacy.util.get_installed_models()
```

**Reply by Vilhelm-Ian** [reply]
it wouldn't be because it would only appear if somebody has it installed they probably know what a spacy model is. I was talking about the option where the user downloads manually spacy and the model

**Reply by mortii** [reply]
@Vilhelm-Ian I have renamed it to this

![Screenshot from 2023-12-19 16-19-55](https://github.com/mortii/anki-morphs/assets/15674619/18549a5f-02ff-4e0f-ae48-133204aea0f2)

**Reply by Vilhelm-Ian** [reply]
would that mean we will deprecate all the other morphemizers

**Reply by mortii** [reply]
@Vilhelm-Ian yes

**Reply by Vilhelm-Ian** [reply]
The problem is to use the builtin morphemizers  is simple. This way users will have to dance with pip. It might scare them off

**Reply by mortii** [reply]
I have always hated the fact that the morphemizers come bundled with the add-on. When they are gone then the addon size goes to 300KB instead of 30MB. 

The built-in morphemizers are simpler to use, but all of them produce worse output than the spacy models. Even if you use a simple whitespace language I think it's worth it to use spacy models because they recognize names.

Supporting only one type of morphemizer also makes the note filter settings simpler.

It will likely scare off many users, but it is still worth it imo.

**Reply by Vilhelm-Ian** [reply]
> The built-in morphemizers are simpler to use, but all of them produce worse output than the spacy models.

This line got me HYPED for the next release

**Comment by mortii** [comment]
Wow, we can potentially implement automatic explanations of words/morphs:
```
spacy.explain("VBZ") returns “verb, 3rd person singular present”.
```
https://spacy.io/usage/linguistic-features#pos-tagging

**Comment by mortii** [comment]
We have to add some custom rules to make hyphens work.
https://spacy.io/usage/linguistic-features#native-tokenizer-additions

```python
import spacy
from spacy.lang.char_classes import ALPHA, ALPHA_LOWER, ALPHA_UPPER
from spacy.lang.char_classes import CONCAT_QUOTES, LIST_ELLIPSES, LIST_ICONS
from spacy.util import compile_infix_regex

# Default tokenizer
nlp = spacy.load("en_core_web_sm")
doc = nlp("mother-in-law")
print([t.text for t in doc]) # ['mother', '-', 'in', '-', 'law']

# Modify tokenizer infix patterns
infixes = (
    LIST_ELLIPSES
    + LIST_ICONS
    + [
        r"(?<=[0-9])[+\\-\\*^](?=[0-9-])",
        r"(?<=[{al}{q}])\\.(?=[{au}{q}])".format(
            al=ALPHA_LOWER, au=ALPHA_UPPER, q=CONCAT_QUOTES
        ),
        r"(?<=[{a}]),(?=[{a}])".format(a=ALPHA),
        # ✅ Commented out regex that splits on hyphens between letters:
        # r"(?<=[{a}])(?:{h})(?=[{a}])".format(a=ALPHA, h=HYPHENS),
        r"(?<=[{a}0-9])[:<>=/](?=[{a}])".format(a=ALPHA),
    ]
)

infix_re = compile_infix_regex(infixes)
nlp.tokenizer.infix_finditer = infix_re.finditer
doc = nlp("mother-in-law")
print([t.text for t in doc]) # ['mother-in-law']
```

**Comment by mortii** [comment]
I now have a basic proof of concept version that uses spacy during recalc, and it recognizes _significantly_ more morphs. It might also turn out to be faster than mecab, we'll see. 

Todo-list before releasing:
- [x] load arbitrary model selected by user
- [x] simplify Morpheme class
- [x] refactor am_db to reflect morpheme class changes
- [x] drop tables if db fails to load on startup
- [x] re-add old morphemizers
    - [x] mecab
    - [x] white space languages
- [x] create spacy wrapper
    - [x] implement hypen rule
- [x] create tests and disable pipes
    - [x] ja
    - [x] nb
    - [x] en
    - [x] de
    - [x] ca (catalan)
    - [x] zh (chinese)
    - [x] hr (croatian)
    - [x] da (danish)
    - [x] nl (dutch)
    - [x] fi (finnish)
    - [x] fr (french)
    - [x] el (greek)
    - [x] it (italian)
    - [x] ko (korean)
    - [x] lt (lithuanian)
    - [x] mk (macedonian)
    - [x] pl (polish)
    - [x] pt (portuguese)
    - [x] ro (romanian)
    - [x] ru (russian)
    - [x] sl (slovenian)
    - [x] es (spanish)
    - [x] sv (swedish)
    - [x] uk (ukrainian)
- [x] redo settings parsing options
    - [x] rename parse -> preprocess 
    - [x] implement with spacy
- [x] Recalc: batch text inputs to nlp (https://spacy.io/usage/processing-pipelines#processing)
    - [x] 53s -> 23s
- [x] update generators to use spacy instead of old morphemizers
    - [x] create tests for generators
    - [x] frequency file generator
        - [x] rename parse -> preprocess
        - [x] implement spacy api
    - [x] readability report generator
        - [x] rename parse -> preprocess
        - [x] implement spacy api
        - [x] rename tabs to numerical & percentage
- [x] Update docs
    - [x] create user guides for installing spacy and spacy models
    - [x] parse has been renamed
        - [x] settings
        - [x] generators

**Comment by Vilhelm-Ian** [comment]
@mortii can you share the current development branch. I really want to try it out

**Reply by mortii** [reply]
Sure, I just have to iron out some kinks first, maybe later today.

**Reply by mortii** [reply]
@Vilhelm-Ian pushed 'spacy' branch now

**Reply by mortii** [reply]
@Vilhelm-Ian have you had the chance to test it yet?

**Comment by mortii** [comment]
Korean produces stuff like this:

w.lemma_: 추대+하+었+다
w.text: 추대했다
w.pos_: VERB

I dont know how we should handle that kind of lemma output...

**Reply by Vilhelm-Ian** [reply]
We should ask "HQ" he is learning korean

**Reply by mortii** [reply]
@HQYang1979 do you know korean? If so, is it ok to show the root of "추대했다" as "추대+하+었+다"? Or should it be something else?

I think the english equivalent would be "promoted" -> "promote"

**Comment by HQYang1979** [comment]
> @HQYang1979 do you know korean? If so, is it ok to show the root of "추대했다" as "추대+하+었+다"? Or should it be something else?

I think the english equivalent would be "promoted" -> "promote"

Should be the same as English showing "promote" only.  The basic form of the said verb is 추대하다. 
w.lemma_: 추대+하+었+다 （I am not sure what does this mean, there is no 추대하었다 or 추대었다）
w.text: 추대했다 (this is the past form)
w.pos_: VERB

**Reply by Vilhelm-Ian** [reply]
> w.lemma_: 추대+하+었+다 （I am not sure what does this mean, there is no 추대하었다 or 추대었다）

spacy sometimes hallucinates.

**Reply by mortii** [reply]
@HQYang1979 Thank you so much! I'll see if I can change it somehow.

**Reply by mortii** [reply]
[This reddit thread](https://www.reddit.com/r/Korean/comments/7oabj9/algorithm_to_get_the_dictionary_form_of_any_word/ds85cxa/) was pretty informative. It seems the first part of the lemma is the stem and the + parts are conjugations.

So for example when I input "눌러", i get:

w.lemma_: 누르+어
w.text: 눌러
w.pos_: NOUN

we can then strip the conjugations from the lemma, and be left with just " 누르"

The downside of this is that we have to make the spacy wrapper more extensive.... but if it means that we can support korean then it is worth it imo.

EDIT:
I can probably just add a new pipe to only the korean nlp object, that way it does not effect the other languages.

**Reply by Vilhelm-Ian** [reply]
I agree

**Reply by mortii** [reply]
I added this pipe to the korean nlp:

```python
@Language.component("stripped_korean_lemma")
def stripped_korean_lemma(doc):
    for w in doc:
        conjugation_position = w.lemma_.find("+")
        if conjugation_position != -1:
            w.lemma_ = w.lemma_[:conjugation_position]
    return doc
```
and now it produces this:

w.lemma_: 누르
w.text: 눌러
w.pos_: NOUN

Pretty clean! I love how adjustable the spacy architecture is.

**Reply by ashprice** [reply]
> w.lemma_: 추대+하+었+다

I am pretty sure this is a morphemic breakdown that might not be obvious to the average language learner/native speaker as opposed to one that is just incorrect. 

I do not know so much about Korean, though, so I don't know of a good source to look to for a breakdown of this kind of analysis. I think you can ask spacy for annotations somehow about this? I don't recall.

**Comment by mortii** [comment]
Chinese (zh) spacy models do not produce lemmas. The spacy devs have discussed this: https://www.github.com/explosion/spaCy/discussions/10386

 I'll create a pipe that sets the lemma = text. This means the U and A in the toolbar will always be the same for chinese.

**Reply by HQYang1979** [reply]
there is no conjugation in Chinese.

**Comment by mortii** [comment]
I have ran into a problem with the pre-trained japanese spacy models, namely that they all use SudachiPy split mode A, which aggressively splits every word/morph as soon as possible, which is really bad.

@ashprice has mentioned this earlier: https://github.com/mortii/anki-morphs/discussions/13#discussioncomment-7291249

The [ginza](https://github.com/megagonlabs/ginza) model fixes this, and it is just as simple to download and use as the other spacy models, _except_ for the fact that it has an arbitrary dependency on an older version of spacy, which makes pip break when trying to install it. This can easily be fixed by just relaxing the spacy dependency that ginza specifies, but the ginza repo is basically unmaintained, so that might be unlikely.  

To fix this, I'm going to  fork ginza and do it myself, and then people can download that model instead.

**Reply by mortii** [reply]
I'm tempted to train my own model instead. Ginza has a lot of baggage.

**Reply by Vilhelm-Ian** [reply]
I think it should be possible with chatgpt AP(you to send the word and in a json to return the pemma and base)I. It shall cost. But I doubt it will be more than 20$

**Reply by mortii** [reply]
@Vilhelm-Ian I'm never never going to spend money on this project, interesting idea though. Spacy has a good tutorial for training compatible models: https://spacy.io/usage/training

**Comment by mortii** [comment]
Bad news: all japanese spacy models (including ginza) split too early. Some examples:
- 本当に -> 本当, に
- 別に -> 別, に
- ごめんなさい -> ごめん, なさい
- 気持ちいい -> 気持ち, いい

They all use sudachi, and the split modes option given to sudachi barely makes a difference.

This means that we cannot replace the current mecab japanese morphemizer with spacy, which means I have to add back the old morphemizers. Not sure how long that will take.

EDIT:
The unidic mecab version also has the same problem.

**Reply by ashprice** [reply]
I think this is a known issue with morphemizers for Japanese...

I don't have a good solution, but I _think_ that https://ichi.moe/ and _know_ that some workflows in corpus linguistics similar to it often solve this by simply matching for the longest string (ie. taking the output and concatenating at the phrasal/word level - not the sentence level). I can't find the Reddit thread where the ichi.moe dev said that now, might be worth reaching out to them in an email. Also, any of the models that include dependency info like Ginza should theoretically allow for a high degree of being able to specify what kind of things you want and do not want to match, but it's going to be at the cost of a huge implementation headache (and I have to say I do not find the Ginza documentation to be very intuitive, I have a few times discovered how to do something interesting or new or involved with it, and then promptly forgot.)

I think it's also how some dictionary tools like Jisho and Yomichan do things as well, but I'm not as sure on that.

I don't know it's a route that one would want to go down, though. It certainly seems a hacky fix, but maybe it would work well; certainly, ichi.moe gets stuff wrong semi-often, but it's maybe more annoying there because it exists as a tool for explanation/reference, whereas with ankimorphs, incorrect results are going to be a lot less annoying - hell, you may not even notice. The important question is the relative one - is it right or wrong more or less often than the alternative choices - not if it is wrong, ever (it surely will be). 

I am pretty sure ichi.moe is right more often than mecab was in the original morphman, and I am pretty sure ichi uses mecab, so maybe it would even improve accuracy for our purposes. But that is just a subjective impression given usage of ichi.moe and anki - even if it led to better output vs. the original morphman, there's no saying that approach will be better elsewhere, for example, that it will improve results relative to the ankimorphs implementation. 

Also, sorry for disappearing for ~2 weeks.

**Reply by ashprice** [reply]
Actually that being viable is kind of contingent on having an appropriate dictionary file. I'll go poke around sudachidict I guess. My suspicion is that you'd have to manually create some dictionary file from some 国語辞典 or something, and at that point, we're reinventing the wheel I guess.

**Reply by mortii** [reply]
@ashprice Right as I wrote that post, I went and took the briefest of glances at the ichi.moe and jisho.org websites for what they use, as they more often then not give good results (better than the built-in mecab anyway) and they both seem to use  JMdict. Like you say, they probably do some fancy work behind the scenes, but it's something I'm tempted to replicate at some point.

**Comment by HQYang1979** [comment]
> EDIT:
The unidic mecab version also has the same problem.

Do you think the Japanese addon that Morphman uses for morphemization is better than Unidic Mecab?

**Reply by mortii** [reply]
@HQYang1979 Yes, it is also in ankimorphs and I have been using it the entire time, so I never realized that the unidic mecab version had this problem.

**Reply by HQYang1979** [reply]
> @HQYang1979 Yes, it is also in ankimorphs and I have been using it the entire time, so I never realized that the unidic mecab version had this problem.

good to know, I thought we can only use mecab. I'll change then

**Reply by mortii** [reply]
@HQYang1979 if you delete the mecab addons then the option: "Japanese ankimorphs" shows up, which is a non-unidic version of mecab

**Comment by mortii** [comment]
The morphemizer option will look like this:
![new-morphemizers](https://github.com/mortii/anki-morphs/assets/15674619/f52141d7-5d37-4f4a-99ef-b077ac8508ac)

The spacy options will only show up after you have installed the models

**Comment by mortii** [comment]
I have a working version now, I just need to test it on windows and macOS, and write the guide for installing spacy and spacy models. Hopefully I can get it done in a day or two.

I also unintentionally managed to make the built-in mecab twice as fast by just cleaning up mecab_wrapper.py a little bit. That might be a sign that I broke something, but it seems to work as intended so hopefully it's just serendipity :man_shrugging:

**Comment by mortii** [comment]
More bad news:
the spaCy installation is getting more complicated by the fact that Anki uses python 3.9, and 'python -m pip install [module]' only installs the module for that particular python version, not all python versions.

So that leaves two options:
1. ~~Have every user install python 3.9~~ 
2. Open anki from terminal where anki will then be used with the python version that user already has installed. 

I think option 2 is probably better since it can be a nightmare to install python 3.9....

EDIT:
option 2 is OS specific so that might not be better. When running the 'anki' command in cmd on windows anki uses the user installed python version, but it does not on ubuntu, so one would have to create a virtual environment, which is probably more of a hassle than installing python 3.9.

EDIT 2:
Actually, you could just install Anki itself from pip: https://betas.ankiweb.net/#via-pypipip so option 2 might work well after all

EDIT 3:
python 3.9 is completely dead, option 1 is is not viable.

**Comment by mortii** [comment]
This is now released in [v0.11](https://github.com/mortii/anki-morphs/releases/tag/v0.11.0-alpha)!
spaCy wasn't the panacea I hoped it would be, nor is it easy to setup, but hopefully someone finds it useful.

**Reply by Vilhelm-Ian** [reply]
thank you very much mortii. I have been waiting for this feature for over a year
